README.md
Disguise-Shadowsocks
====================
Disguise-Shadowsocks is based on Shadowsocks, and keeps the TCP stream safer from GFW.
#Principle:
>Data transported between Shadowsocks client and server is raw TCP stream in the eye of GFW, that is really not decryptable.
>But if you think GFW will let go any raw TCP stream, you are wrong.
>Many Chinese friends have found that the port or even IP of their Shadowsocks serve is blocked after a while of using.
>So I can rationally suppose that GFW is watching all TCP stream and recording those it can't parse, once its doubt to some address exceeds a given limit, it will interfere or even block that suspicious address.
>So what do we do? One feasible idea is to eliminate your suspicious characters in your TCP stream, right?
>How to? I think I can disguise all TCP data as HTTP which is parsable to GFW, and make GFW think me a good guy.
>
>This is what Disguise-Shadowsocks does: client sends HTTP request to server, in whose DATA part is the real TCP placed. In the same way, server replys the disguised TCP data as HTTP response.

By the way, original Shadowsocks client is still possible to connect to Disguise-Shadowsocks for compatibility concerning.
#Usage:
>Notice an extra config option "udp_server_port" (-u for command line) than Shadowsocks, it allows you to specify port for udp server.
Other configuration remains the same with Shadowsocks.

Disguise-Shadowsocks works fine for me now. If any problem, launch new issue please.

CHANGES
2016-1-1
Compatible with orginal shadowsocks

2015-11-26
Now, the fake http request header is only added before the first tcp segment from ssclient to ssserver
the fake http response header is only added before the first tcp segment from ssserver to ssclient
it works on any vps now!

2015-07-29
Added disguise function against shadowsocks.
All tcp data will appear just like http request and response to avoid censorship from GFW.
Split tcp server and udp server on different ports to avoid censorship which gets smarter nowadays.

2.6.11 2015-07-10
- Fix a compatibility issue in UDP Relay

2.6.10 2015-06-08
- Optimize LRU cache
- Refine logging

2.6.9 2015-05-19
- Fix a stability issue on Windows

2.6.8 2015-02-10
- Support multiple server ip on client side
- Support --version
- Minor fixes

2.6.7 2015-02-02
- Support --user
- Support CIDR format in --forbidden-ip
- Minor fixes

2.6.6 2015-01-23
- Fix a crash in forbidden list

2.6.5 2015-01-18
- Try both 32 bit and 64 bit dll on Windows

2.6.4 2015-01-14
- Also search lib* when searching libraries

2.6.3 2015-01-12
- Support --forbidden-ip to ban some IP, i.e. localhost
- Search OpenSSL and libsodium harder
- Now works on OpenWRT

2.6.2 2015-01-03
- Log client IP

2.6.1 2014-12-26
- Fix a problem with TCP Fast Open on local side
- Fix sometimes daemon_start returns wrong exit status

2.6 2014-12-21
- Add daemon support

2.5 2014-12-11
- Add salsa20 and chacha20

2.4.3 2014-11-10
- Fix an issue on Python 3
- Fix an issue with IPv6

2.4.2 2014-11-06
- Fix command line arguments on Python 3
- Support table on Python 3
- Fix TCP Fast Open on Python 3

2.4.1 2014-11-01
- Fix setup.py for non-utf8 locales on Python 3

2.4 2014-11-01
- Python 3 support
- Performance improvement
- Fix LRU cache behavior

2.3.2 2014-10-11
- Fix OpenSSL on Windows

2.3.1 2014-10-09
- Does not require M2Crypto any more

2.3 2014-09-23
- Support CFB1, CFB8 and CTR mode of AES
- Do not require password config when using port_password
- Use SIGTERM instead of SIGQUIT on Windows

2.2.2 2014-09-14
- Fix when multiple DNS set, IPv6 only sites are broken

2.2.1 2014-09-10
- Support graceful shutdown
- Fix some bugs

2.2.0 2014-09-09
- Add RC4-MD5 encryption

2.1.0 2014-08-10
- Use only IPv4 DNS server
- Does not ship config.json
- Better error message

2.0.12 2014-07-26
- Support -q quiet mode
- Exit 0 when showing help with -h

2.0.11 2014-07-12
- Prefers IP addresses over hostnames, more friendly with socksify and openvpn

2.0.10 2014-07-11
- Fix UDP on local

2.0.9 2014-07-06
- Fix EWOULDBLOCK on Windows
- Fix Unicode config problem on some platforms

2.0.8 2014-06-23
- Use multiple DNS to query hostnames

2.0.7 2014-06-21
- Fix fastopen on local
- Fallback when fastopen is not available
- Add verbose logging mode -vv
- Verify if hostname is valid

2.0.6 2014-06-19
- Fix CPU 100% on POLL_HUP
- More friendly logging

2.0.5 2014-06-18
- Support a simple config format for multiple ports

2.0.4 2014-06-12
- Fix worker master

2.0.3 2014-06-11
- Fix table encryption with UDP

2.0.2 2014-06-11
- Add asynchronous DNS in TCP relay

2.0.1 2014-06-05
- Better logging
- Maybe fix bad file descriptor

2.0 2014-06-05
- Use a new event model
- Remove gevent
- Refuse to use default password
- Fix a problem when using multiple passwords with table encryption

1.4.5 2014-05-24
- Add timeout in TCP server
- Close sockets in master process

1.4.4 2014-05-17
- Support multiple workers

1.4.3 2014-05-13
- Fix Windows

1.4.2 2014-05-10
- Add salsa20-ctr cipher

1.4.1 2014-05-03
- Fix error log
- Fix EINPROGESS with some version of gevent

1.4.0 2014-05-02
- Adds UDP relay
- TCP fast open support on Linux 3.7+

1.3.7 2014-04-10
- Fix a typo in help

1.3.6 2014-04-10
- Fix a typo in help

1.3.5 2014-04-07
- Add help
- Change default local binding address into 127.0.0.1

1.3.4 2014-02-17
- Fix a bug when no config file exists
- Client now support multiple server ports and multiple server/port pairs
- Better error message with bad config.json format and wrong password

1.3.3 2013-07-09
- Fix default key length of rc2

1.3.2 2013-07-04
- Server will listen at server IP specified in config
- Check config file and show some warning messages

1.3.1 2013-06-29
- Fix -c arg

1.3.0 2013-06-22
- Move to pypi

1.2.3 2013-06-14
- add bind address

1.2.2 2013-05-31
- local can listen at ::0 with -6 arg; bump 1.2.2

1.2.1 2013-05-23
- Fix an OpenSSL crash

1.2 2013-05-22
- Use random iv, we finally have strong encryption

1.1.1 2013-05-21
- Add encryption, AES, blowfish, etc.

1.1 2013-05-16
- Support IPv6 addresses (type 4)
- Drop Python 2.5 support

1.0 2013-04-03
- Fix -6 IPv6

0.9.4 2013-03-04
- Support Python 2.5

0.9.3 2013-01-14
- Fix conn termination null data

0.9.2 2013-01-05
- Change default timeout

0.9.1 2013-01-05
- Add Travis-CI test

0.9 2012-12-30
- Replace send with sendall, fix FreeBSD

0.6 2012-12-06
- Support args

0.5 2012-11-08
- Fix encryption with negative md5sum

0.4 2012-11-02
- Move config into a JSON file
- Auto-detect config path

0.3 2012-06-06
- Move socks5 negotiation to local

0.2 2012-05-11
- Add -6 arg for IPv6
- Fix socket.error

0.1 2012-04-20
- Initial version


shadowsocks/socks5.py
#coding=utf-8
#copyright by xiaoxia
#http://xiaoxia.org/2011/03/29/written-by-python-socks5-server/
import socket, sys, select, SocketServer, struct, time

class ThreadingTCPServer(SocketServer.ThreadingMixIn, SocketServer.TCPServer): pass
class Socks5Server(SocketServer.StreamRequestHandler):
    def handle_tcp(self, sock, remote):
        fdset = [sock, remote]
        while True:
            r, w, e = select.select(fdset, [], [])
            if sock in r:
                if remote.send(sock.recv(4096)) <= 0: break
            if remote in r:
                if sock.send(remote.recv(4096)) <= 0: break
    def handle(self):
        try:
            print 'socks connection from ', self.client_address
            sock = self.connection
            # 1. Version
            sock.recv(262)
            sock.send(b"\x05\x00");
            # 2. Request
            data = self.rfile.read(4)
            mode = ord(data[1])
            addrtype = ord(data[3])
            if addrtype == 1:       # IPv4
                addr = socket.inet_ntoa(self.rfile.read(4))
            elif addrtype == 3:     # Domain name
                addr = self.rfile.read(ord(sock.recv(1)[0]))
            port = struct.unpack('>H', self.rfile.read(2))
            reply = b"\x05\x00\x00\x01"
            try:
                if mode == 1:  # 1. Tcp connect
                    remote = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
                    remote.connect((addr, port[0]))
                    print 'Tcp connect to', addr, port[0]
                else:
                    reply = b"\x05\x07\x00\x01" # Command not supported
                local = remote.getsockname()
                reply += socket.inet_aton(local[0]) + struct.pack(">H", local[1])
            except socket.error:
                # Connection refused
                reply = '\x05\x05\x00\x01\x00\x00\x00\x00\x00\x00'
            sock.send(reply)
            # 3. Transfering
            if reply[1] == '\x00':  # Success
                if mode == 1:    # 1. Tcp connect
                    self.handle_tcp(sock, remote)
        except socket.error:
            print 'socket error'
def main():
    server = ThreadingTCPServer(('', 1555), Socks5Server)
    server.serve_forever()
if __name__ == '__main__':
    main()


shadowsocks/disguise.py
#coding=utf-8
__author__ = 'easton'
import unittest, os, random, string, re

def id_generator(size=10, chars=string.ascii_uppercase + string.digits):
    return ''.join(random.choice(chars) for _ in range(size))

'''
in shadowsocks, sometimes multiple small packages are sent, but one big packages received
so, the regular expression in the extraction should not start in the beginning(^)
'''

http_request_payload_template = \
'''POST /{0}.png HTTP/1.1\r
Host: bighouse.com\r
Connection: keep-alive\r
Pragma: no-cache\r
Cache-Control: no-cache\r
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8\r
Upgrade-Insecure-Requests: 1\r
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/47.0.2526.73 Safari/537.36\r
Content-Type: application/octet-stream\r
DNT: 1\r
Accept-Encoding: gzip, deflate\r
Accept-Language: en-US,en;q=0.8,zh-CN;q=0.6,zh;q=0.4,zh-TW;q=0.2\r
\r
{1}'''

# easton: removed the tail for simplicity, now we only need to add header before the first tcp segment,
# but don't need to append tail after the last
http_responce_payload_template = '''HTTP/1.1 200 OK\r
Content-Type: image/png\r
Connection: keep-alive\r
\r
\x89PNG\r\n\x1a\xea{0}'''
#\x89PNG\r\n\x1a\xea{0}\xea\x00\x00\x00IEND\xaeB\x60\x82'''

disguise_count = 0
extract_count = 0
extract_success = 0

def replace_LF_with_CRLF(string):
    return string.replace('\n', '\r\n')

def disguise_as_http_request(data):
    global disguise_count
    disguise_count += 1
    #print 'disguising', disguise_count
    return (
        # comfirmed: '{}'.format can handle binary
        http_request_payload_template.format(id_generator(), data))

def extract_from_fake_http_request(request_str):
    """
    :type request_str: bytearray
    """
    global extract_count, extract_success
    extract_count += 1
    #to_strip = '(GET.*?previous_pic=|; \r\n\r\n)'
    to_strip = '(POST.*?q=0\.2\r\n\r\n)'
    is_disguised = False
    #'dont match the beginning'
    if re.search(to_strip, request_str, flags=re.DOTALL):
        is_disguised = True
        extract_success += 1
    striped = re.sub(to_strip, '', request_str, flags=re.DOTALL)
    return striped, is_disguised

def disguise_as_http_responce(data):
    global disguise_count
    disguise_count += 1
    return http_responce_payload_template.format(data)

def extract_from_fake_http_responce(responce_str):
    global extract_count, extract_success
    extract_count += 1
    'dont match the beginning'
    if re.search('(HTTP.*?\x89PNG\r\n\x1a\xea|\xea\x00\x00\x00IEND\xaeB\x60\x82)', responce_str, flags=re.DOTALL):
        extract_success += 1
    return str(re.sub('(HTTP.*?\x89PNG\r\n\x1a\xea|\xea\x00\x00\x00IEND\xaeB\x60\x82)', '', responce_str, flags=re.DOTALL))

class Test(unittest.TestCase):
    def est_replace_LF_with_CRLF(self):
        for i in bytearray(replace_LF_with_CRLF(http_responce_payload_template)):
            print hex(i),

    def est_disguise_as_http_request(self):
        ba = bytearray((disguise_as_http_request('\x88ciphertext\n')))
        for i in ba:
            print hex(i),
        print
        print str(ba)

    def est_disguise_as_http_responce(self):
        for i in bytearray(disguise_as_http_responce('ciphertext\n')):
            print hex(i),

    def est_extract_from_fake_http_request(self):
        print extract_from_fake_http_request(disguise_as_http_request('\x99'))

    def est_extract_from_fake_http_responce(self):
        print repr(extract_from_fake_http_responce(disguise_as_http_responce('\x99')))

    def test_bytearray_replace(self):
        ba = bytearray('get\r\n=\x98\x50\r\n')
        print repr(re.sub('(^g.*?=|\r\n$)', '', ba, flags=re.DOTALL))

    def est_str_replace(self):
        print '{}'.replace('{}', '\x41')

    def test_str_format_with_binary(self):
        print '{}'.format('\xff')

    def est_unicode_reg(self):
        print len(u'\x88aa'.encode('utf-8'))
        #print repr(re.search(u'(\x88)', u'\x88adfasdf').group(1))

if __file__ == '__main__':
    unittest.main()

shadowsocks/shell.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
#
# Copyright 2015 clowwindy
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may
# not use this file except in compliance with the License. You may obtain
# a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations
# under the License.

from __future__ import absolute_import, division, print_function, \
    with_statement

import os
import json
import sys
import getopt
import logging
from shadowsocks.common import to_bytes, to_str, IPNetwork
from shadowsocks import encrypt


VERBOSE_LEVEL = 5

verbose = 0


def check_python():
    info = sys.version_info
    if info[0] == 2 and not info[1] >= 6:
        print('Python 2.6+ required')
        sys.exit(1)
    elif info[0] == 3 and not info[1] >= 3:
        print('Python 3.3+ required')
        sys.exit(1)
    elif info[0] not in [2, 3]:
        print('Python version not supported')
        sys.exit(1)


def print_exception(e):
    global verbose
    logging.error(e)
    if verbose > 0:
        import traceback
        traceback.print_exc()


def print_shadowsocks():
    version = ''
    try:
        import pkg_resources
        version = pkg_resources.get_distribution('shadowsocks').version
    except Exception:
        pass
    print('Shadowsocks %s' % version)


def find_config():
    config_path = 'config.json'
    if os.path.exists(config_path):
        return config_path
    config_path = os.path.join(os.path.dirname(__file__), '../', 'config.json')
    if os.path.exists(config_path):
        return config_path
    return None


def check_config(config, is_local):
    if config.get('daemon', None) == 'stop':
        # no need to specify configuration for daemon stop
        return

    if is_local and not config.get('password', None):
        logging.error('password not specified')
        print_help(is_local)
        sys.exit(2)

    if not is_local and not config.get('password', None) \
            and not config.get('port_password', None):
        logging.error('password or port_password not specified')
        print_help(is_local)
        sys.exit(2)

    if 'local_port' in config:
        config['local_port'] = int(config['local_port'])

    if 'server_port' in config and type(config['server_port']) != list:
        config['server_port'] = int(config['server_port'])

    if config.get('local_address', '') in [b'0.0.0.0']:
        logging.warn('warning: local set to listen on 0.0.0.0, it\'s not safe')
    if config.get('server', '') in ['127.0.0.1', 'localhost']:
        logging.warn('warning: server set to listen on %s:%s, are you sure?' %
                     (to_str(config['server']), config['server_port']))
    if (config.get('method', '') or '').lower() == 'table':
        logging.warn('warning: table is not safe; please use a safer cipher, '
                     'like AES-256-CFB')
    if (config.get('method', '') or '').lower() == 'rc4':
        logging.warn('warning: RC4 is not safe; please use a safer cipher, '
                     'like AES-256-CFB')
    if config.get('timeout', 300) < 100:
        logging.warn('warning: your timeout %d seems too short' %
                     int(config.get('timeout')))
    if config.get('timeout', 300) > 600:
        logging.warn('warning: your timeout %d seems too long' %
                     int(config.get('timeout')))
    if config.get('password') in [b'mypassword']:
        logging.error('DON\'T USE DEFAULT PASSWORD! Please change it in your '
                      'config.json!')
        sys.exit(1)
    if config.get('user', None) is not None:
        if os.name != 'posix':
            logging.error('user can be used only on Unix')
            sys.exit(1)

    encrypt.try_cipher(config['password'], config['method'])


def get_config(is_local):
    global verbose

    logging.basicConfig(level=logging.INFO,
                        format='%(levelname)-s: %(message)s')
    if is_local:
        shortopts = 'hd:s:b:p:u:k:l:m:c:t:vq'
        longopts = ['help', 'fast-open', 'pid-file=', 'log-file=', 'user=',
                    'version']
    else:
        shortopts = 'hd:s:p:u:k:m:c:t:vq'
        longopts = ['help', 'fast-open', 'pid-file=', 'log-file=', 'workers=',
                    'forbidden-ip=', 'user=', 'version']
    try:
        config_path = find_config()
        optlist, args = getopt.getopt(sys.argv[1:], shortopts, longopts)
        for key, value in optlist:
            if key == '-c':
                config_path = value

        if config_path:
            logging.info('loading config from %s' % config_path)
            with open(config_path, 'rb') as f:
                try:
                    config = json.loads(f.read().decode('utf8'),
                                        object_hook=_decode_dict)
                except ValueError as e:
                    logging.error('found an error in config.json: %s',
                                  e.message)
                    sys.exit(1)
        else:
            config = {}

        v_count = 0
        for key, value in optlist:
            if key == '-p':
                config['server_port'] = int(value)
            elif key == '-u':
                config['udp_server_port'] = int(value)
            elif key == '-k':
                config['password'] = to_bytes(value)
            elif key == '-l':
                config['local_port'] = int(value)
            elif key == '-s':
                config['server'] = to_str(value)
            elif key == '-m':
                config['method'] = to_str(value)
            elif key == '-b':
                config['local_address'] = to_str(value)
            elif key == '-v':
                v_count += 1
                # '-vv' turns on more verbose mode
                config['verbose'] = v_count
            elif key == '-t':
                config['timeout'] = int(value)
            elif key == '--fast-open':
                config['fast_open'] = True
            elif key == '--workers':
                config['workers'] = int(value)
            elif key == '--user':
                config['user'] = to_str(value)
            elif key == '--forbidden-ip':
                config['forbidden_ip'] = to_str(value).split(',')
            elif key in ('-h', '--help'):
                if is_local:
                    print_local_help()
                else:
                    print_server_help()
                sys.exit(0)
            elif key == '--version':
                print_shadowsocks()
                sys.exit(0)
            elif key == '-d':
                config['daemon'] = to_str(value)
            elif key == '--pid-file':
                config['pid-file'] = to_str(value)
            elif key == '--log-file':
                config['log-file'] = to_str(value)
            elif key == '-q':
                v_count -= 1
                config['verbose'] = v_count
    except getopt.GetoptError as e:
        print(e, file=sys.stderr)
        print_help(is_local)
        sys.exit(2)

    if not config:
        logging.error('config not specified')
        print_help(is_local)
        sys.exit(2)

    config['password'] = to_bytes(config.get('password', b''))
    config['method'] = to_str(config.get('method', 'aes-256-cfb'))
    config['port_password'] = config.get('port_password', None)
    config['timeout'] = int(config.get('timeout', 300))
    config['fast_open'] = config.get('fast_open', False)
    config['workers'] = config.get('workers', 1)
    config['pid-file'] = config.get('pid-file', '/var/run/shadowsocks.pid')
    config['log-file'] = config.get('log-file', '/var/log/shadowsocks.log')
    config['verbose'] = config.get('verbose', False)
    config['local_address'] = to_str(config.get('local_address', '127.0.0.1'))
    config['local_port'] = config.get('local_port', 1080)
    if is_local:
        if config.get('server', None) is None:
            logging.error('server addr not specified')
            print_local_help()
            sys.exit(2)
        else:
            config['server'] = to_str(config['server'])
    else:
        config['server'] = to_str(config.get('server', '0.0.0.0'))
        try:
            config['forbidden_ip'] = \
                IPNetwork(config.get('forbidden_ip', '127.0.0.0/8,::1/128'))
        except Exception as e:
            logging.error(e)
            sys.exit(2)
    config['server_port'] = config.get('server_port', 8388)

    logging.getLogger('').handlers = []
    logging.addLevelName(VERBOSE_LEVEL, 'VERBOSE')
    if config['verbose'] >= 2:
        level = VERBOSE_LEVEL
    elif config['verbose'] == 1:
        level = logging.DEBUG
    elif config['verbose'] == -1:
        level = logging.WARN
    elif config['verbose'] <= -2:
        level = logging.ERROR
    else:
        level = logging.INFO
    verbose = config['verbose']
    logging.basicConfig(level=level,
                        format='%(asctime)s %(levelname)-8s %(message)s',
                        datefmt='%Y-%m-%d %H:%M:%S')

    check_config(config, is_local)

    return config


def print_help(is_local):
    if is_local:
        print_local_help()
    else:
        print_server_help()


def print_local_help():
    print('''usage: sslocal [OPTION]...
A fast tunnel proxy that helps you bypass firewalls.

You can supply configurations via either config file or command line arguments.

Proxy options:
  -c CONFIG              path to config file
  -s SERVER_ADDR         server address
  -p SERVER_PORT         server port, default: 8388
  -u udp_server_port
  -b LOCAL_ADDR          local binding address, default: 127.0.0.1
  -l LOCAL_PORT          local port, default: 1080
  -k PASSWORD            password
  -m METHOD              encryption method, default: aes-256-cfb
  -t TIMEOUT             timeout in seconds, default: 300
  --fast-open            use TCP_FASTOPEN, requires Linux 3.7+

General options:
  -h, --help             show this help message and exit
  -d start/stop/restart  daemon mode
  --pid-file PID_FILE    pid file for daemon mode
  --log-file LOG_FILE    log file for daemon mode
  --user USER            username to run as
  -v, -vv                verbose mode
  -q, -qq                quiet mode, only show warnings/errors
  --version              show version information

Online help: <https://github.com/shadowsocks/shadowsocks>
''')


def print_server_help():
    print('''usage: ssserver [OPTION]...
A fast tunnel proxy that helps you bypass firewalls.

You can supply configurations via either config file or command line arguments.

Proxy options:
  -c CONFIG              path to config file
  -s SERVER_ADDR         server address, default: 0.0.0.0
  -p SERVER_PORT         server port, default: 8388
  -u udp_server_port
  -k PASSWORD            password
  -m METHOD              encryption method, default: aes-256-cfb
  -t TIMEOUT             timeout in seconds, default: 300
  --fast-open            use TCP_FASTOPEN, requires Linux 3.7+
  --workers WORKERS      number of workers, available on Unix/Linux
  --forbidden-ip IPLIST  comma seperated IP list forbidden to connect

General options:
  -h, --help             show this help message and exit
  -d start/stop/restart  daemon mode
  --pid-file PID_FILE    pid file for daemon mode
  --log-file LOG_FILE    log file for daemon mode
  --user USER            username to run as
  -v, -vv                verbose mode
  -q, -qq                quiet mode, only show warnings/errors
  --version              show version information

Online help: <https://github.com/shadowsocks/shadowsocks>
''')


def _decode_list(data):
    rv = []
    for item in data:
        if hasattr(item, 'encode'):
            item = item.encode('utf-8')
        elif isinstance(item, list):
            item = _decode_list(item)
        elif isinstance(item, dict):
            item = _decode_dict(item)
        rv.append(item)
    return rv


def _decode_dict(data):
    rv = {}
    for key, value in data.items():
        if hasattr(value, 'encode'):
            value = value.encode('utf-8')
        elif isinstance(value, list):
            value = _decode_list(value)
        elif isinstance(value, dict):
            value = _decode_dict(value)
        rv[key] = value
    return rv
